src/hotspot/share/opto/memnode.cpp

```diff
@@ -1163,13 +1163,39 @@ Node* MemNode::can_see_stored_value(Node* st, PhaseValues* phase) const {
       if (store_Opcode() != st->Opcode()) {
         return nullptr;
       }
-      // LoadVector/StoreVector needs additional check to ensure the types match.
+      // LoadVector/StoreVector need additional checks
       if (st->is_StoreVector()) {
+        // Ensure that types match
         const TypeVect*  in_vt = st->as_StoreVector()->vect_type();
         const TypeVect* out_vt = as_LoadVector()->vect_type();
         if (in_vt != out_vt) {
           return nullptr;
         }
```

Comment:
I see there is a vector type check here. Do we not need that for the code in `StoreNode::Identity`? "Normal" stores like `StoreB` and `StoreI` have the type implicit, but for vector nodes, this type is hidden in the `vect_type()`, so I suspect you need to check it.

I imagine a scenario where we store a float-vector, and then read from the same address as int-vector. Is that ok, or would we need a ReinterpretCast node? I'm not sure, but it would be worth trying to create some tests to check that.

Reply:
Can you actually do that: store a float-vector to an int-array? Or is that maybe only possible with Unsafe somehow? Or maybe completely impossible?

Reply:
The API doesn't seem to allow this. I'm not sure of what one could do with Unsafe. Would we have to take potential Unsafe operations into account?